### PR TITLE
engine: introduce SstWriterBuilder

### PR DESCRIPTION
**Pull request by overvenus**
## What have you changed? (mandatory)

Introduce SstWriterBuilder, it builds a SstWriter which support building an in-memory SST.

For now, no one uses in-memory sst, it is added for backup, it helps backup increasing throughput by not performing any IO operation.

## What are the type of the changes? (mandatory)

- New feature (change which adds functionality)

## How has this PR been tested? (mandatory)

Unit tests.

## Does this PR affect documentation (docs) or release note? (mandatory)

No.

## Does this PR affect tidb-ansible update? (mandatory)

No.

### CODE REVIEW

File: components/engine/src/rocks/sst.rs [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::rocks::util;
+    use tempfile::Builder;
+
+    #[test]
+    fn test_somke() {
```
Comment (Connor1996) [inline components/engine/src/rocks/sst.rs]: What is `somke`?

Comment (kennytm) [reply components/engine/src/rocks/sst.rs]: smoke 🤣

**Comment by siddontang** [comment]
why do we need to build a memory SST? 
when can we use it?

File: components/engine/src/rocks/sst.rs [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+    }
+
+    /// Builder a SstWriter.
+    pub fn build(self, path: &str) -> Result<SstWriter, String> {
+        let mut env = None;
+        let mut io_options = if let Some(db) = self.db.as_ref() {
+            env = db.env();
+            let handle = db
+                .cf_handle(self.cf.unwrap_or(CF_DEFAULT))
+                .map_or_else(|| Err(format!("CF {:?} is not found", self.cf)), Ok)?;
```
Comment (kennytm) [inline components/engine/src/rocks/sst.rs]: 😕 If `cf_handle` returns a `Result` you could use `map_err` here, if `cf_handle` returns an `Option` you could use `ok_or_else` here

File: components/engine/src/rocks/sst.rs (hunk 1 above)
Comment (siddontang) [reply components/engine/src/rocks/sst.rs]: what is smoke?

**Comment by overvenus** [comment]
@siddontang Good questions! Answered in the PR description.

File: components/engine/src/rocks/sst.rs (hunk 1 above)
Comment (overvenus) [reply components/engine/src/rocks/sst.rs]: It's smoke testing. https://en.wikipedia.org/wiki/Smoke_testing_(software)

Comment (kennytm) [reply components/engine/src/rocks/sst.rs]: <https://en.wikipedia.org/wiki/Smoke_testing_(software)>

**Pull request merged**